refactor: actually make the package usable for consumers

napei opened the pull request:
This PR implements a new UUID generator for requests, removing an erroneous dependency on the `uuid` package which was not available for client scripts. 

Along with this, are development configuration updates and code formatting to standardise the repo.

TasoOneAsia commented:
Thanks for the PR,, there was to be more activity in this repo but I have found myself quite busy :/

TasoOneAsia commented on src/server/sv_utils.ts:
```diff
@@ -55,6 +53,17 @@ export class ServerUtils {
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  private uuidV4(options?: any, buf?: any, offset?: any): string {
+    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function (c) {
+      const r = (Math.random() * 16) | 0,
+        v = c == "x" ? r : (r & 0x3) | 0x8;
+      return v.toString(16);
+    });
+  }
+
```
Honestly considering this is probably one of the slowest parts of execution, we may as well reduce the uuid length as a chance of collision during promise resolution is somewhat low, maybe something like a length of 4, but I am open to thoughts. This works fine though.

napei replied:
I added this because it's currently broken, as uuid isn't a dependency and it doesn't compile nicely for node in cfx as it doesn't support `require` (hence the change to esnext).

Doesn't have to be `uuid` does it?
Could do something like this (used in another resource already)

```js
/**
 * Creates a unique ID, such as `notif-0-2ammqh9cevf`
 *
 * @return {*} Unique ID
 */
function uid(prefix) {
  return `${prefix.trim()}-${(idCounter++).toString(36)}-${Math.floor(Math.random() * Number.MAX_SAFE_INTEGER).toString(36)}`;
}

```

napei replied:
> a chance of collision during promise resolution is somewhat low

I don't know about this library. But I've had promise collision many many times with `qb-core`'s `TriggerCallback which is why I'm here. Might not be as low as you think, especially if you have 2 resources calling the same export at the same time (on startup).

TasoOneAsia reviewed (approved):
Small thing to consider but this can be merged.

napei commented:
> Thanks for the PR,, there was to be more activity in this repo but I have found myself quite busy :/

No worries. Was doing this live with `chip` on discord so apologies for zero context. Build is broken for me on FiveM as the `uuid` library isn't a dependency and compling it requires `require("crypto")` which doesn't seem to be available even though it's Node.

TasoOneAsia commented:
Yes, since we only have v8 on client, this previous implementation was broken. I'm on my phone right now, but a fix exists in NPWD. It may be useful to keep consistent with that as 

I eventually want to move directly to using this as a dep package in NPWD rather than having its own implementation.

napei commented:
> Yes, since we only have v8 on client, this previous implementation was broken. I'm on my phone right now, but a fix exists in NPWD. It may be useful to keep consistent with that as
> 
> I eventually want to move directly to using this as a dep package in NPWD rather than having its own implementation.

I think you're referring to this. In that case, could just move that implementation into this package, as an exported helper for both client and server.

Tbh, I would argue if you really need a full-blown uuidv4 anyway (https://github.com/project-error/pe-utils/pull/68#discussion_r786328639)?

https://github.com/project-error/npwd/blob/0db0e3bb30bbcd7a406102869833f5b6d3970be4/resources/utils/fivem.ts#L4-L28

TasoOneAsia commented:
Yea I don't think spec compliance is needed here but I just needed a UUID that wouldn't collide while waiting for the promise to resolve.

I'm open to any solution really that retains behavior.

TasoOneAsia commented:
Honestly could even be a simple id tied to the promise request, that increments on promise event init.

napei commented:
> Yea I don't think spec compliance is needed here but I just needed a UUID that wouldn't collide while waiting for the promise to resolve.
> 
> I'm open to any solution really that retains behavior.

See 7fda179. I believe this is fine. It's random enough between `0` and `Number.MAX_SAFE_INTEGER` and then to add to that there's a counter in the instance of `ServerUtilities` that will ensure there are less conflicts.

This way, you're calling `Math.random` once, instead of 32 times.

I also pushed ffb157b just to simplify the tsconfig a little, as there were some double-ups in defining things.

napei commented:
I've also pushed 3c52be2 which fixes the master CI action, as that had a syntax error and was emailing me that it failed on every push.

TasoOneAsia reviewed (approved):
I'm a fan of this id implementation, would you mind also changing the `cl_utils.ts` implementation to call the same thing? I can add it, in the morning, if you haven't gotten around to it.

napei commented:
@TasoOneAsia I've pushed 640135d which adds a shared helper function for generating a prefixed UUID, and updated the client implementation to follow.

I also pushed 4d01af2 which just runs prettier and adds a `format` script to `package.json`.

I do apologise for the scope of this PR. Sometimes I can't help myself.

TasoOneAsia commented:
> I do apologise for the rather broad scope of this PR. Sometimes I can't help myself.

No worries :), I am happy somebody is taking the time to iron out some of the pending work left with this package.  I'll merge this tomorrow morning hopefully.

napei commented:
> > I do apologise for the rather broad scope of this PR. Sometimes I can't help myself.
> 
> No worries :), I am happy somebody is taking the time to iron out some of the pending work left with this package. I'll merge this tomorrow morning hopefully.

While I'm in here, anything else you want looked at? 😂

TasoOneAsia commented:
> While I'm in here, anything else you want looked at? 😂

Beware what you ask for 😄, there are some chores around the project that need work but can be a little annoying to solve. 

One being the current build process, the nature of FiveM's type packages being (`@citizenfx/client`, `@citizenfx/server`) means that creating a single entry point for both scopes (as the package currently operates) will generate TS compilation errors due to type collisions. 

This makes bundling both sides in one package a pain for automation. One solution would be to split into a `server`/`client` package but I am not particularly a fan of adding unnecessary scoped packages to npm.

If you are interested in contributing, our Discord may be the best place for casual discussion.

napei commented:
Okay, as per talking on discord, here's a bunch of changes to bring the library up to scratch.

- Added a `tsconfig.json` file into `src` to aid in building.
- Leveraged [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) and `tsc --build` to make this build a **lot** easier as one cohesive project. 
- Removed redundant docs build script. You can just run `yarn docs:gen` now and `yarn docs:server` to preview locally
- Added prettier plugins for `eslint` and updated configs to be more in line with modern linting (See [gts](https://github.com/google/gts) for reference)
- Consolidated the `common/` folder (was previously `misc/`). `common/types.ts` may need to be looked at later to further consolidate and neaten it. As the project is compiled to a single entrypoint, then it makes more sense to move these typedefs closer to where they are used i.e. inside `client/` and `server/`

Hopefully this is all okay. Please review and test to double check it's all good.

TasoOneAsia commented on .prettierrc.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "trailingComma": "es5",
+  "arrowParens": "avoid",
+  "printWidth": 140
```
Do I smell an ultra wide monitor here? Mind making this 110, only because I like tiling my windows next to each other and sadly don't have an ultra wide.
```suggestion
  "printWidth": 110
```

napei replied:
Ah sorry yeah 2K monitor haha. I'll change and recommit it all.

napei replied:
Funnily enough, this didn't change any code.

TasoOneAsia commented on .github/workflows/master-ci.yaml:
```diff
@@ -15,14 +15,14 @@ jobs:
         uses: actions/setup-node@v1
         with:
-          node-version: '14.x'
+          node-version: "12.x" # this is the version of citizenfx/node
```
FXServer builds released after artifact 4550, should be running Node 16 on the server now.  Although in a build context, it doesn't really matter.

TasoOneAsia commented on .github/workflows/publish.yml:
```diff
@@ -78,12 +78,10 @@ jobs:
         uses: actions/setup-node@v1
         with:
-          node-version: '14.x'
+          node-version: "12.x"
```
Whole ecosystem is moving towards latest LTS so we should also keep up to date and match with node 16.x when possible.

TasoOneAsia commented:
Scope of changes are a lot to review, so I am just going to leave a couple of preliminary comments while I continue reading.

TasoOneAsia reviewed (commented):
Still need to finish full code review but here.

TasoOneAsia commented on src/client/functions.ts:
```diff
@@ -7,17 +7,7 @@ type CallbackFn<T> = (data: T, cb: NuiCallbackFunc) => void;
  *  @param callback - The callback function
  */
-export const RegisterNuiCB = <T = any>(event: string, callback: CallbackFn<T>): void => {
+export const RegisterNuiCB = <T = unknown>(event: string, callback: CallbackFn<T>): void => {
```
> <T = unknown>  

Stricter, yes, but also less forgiving to beginners with no experience in using generics. Lets just keep it as any, and anyone who wants strict typing can just use the generic.

napei replied:
Fixed in 8c93f02

TasoOneAsia reviewed (approved):
Actually this PR wasn't bad at all to review, thought it would have been much longer based on GH's description, but  `yarn.lock`  + `docgen` at it again. From overview of the logic, it looks sound enough.

merge pending final testing + adding NPM package publishing into automation and CI now that builds are less-scuffed

napei commented:
> Actually this PR wasn't bad at all to review, thought it would have been much longer based on GH's description, but `yarn.lock` + `docgen` at it again. From overview of the logic, it looks sound enough.
> 
> merge pending final testing + adding NPM package publishing into automation and CI now that builds are less-scuffed

Well that's good, appreciate the review. I've tested the net promise functionality (which was the whole original point of this PR) in a server and that's working fine. I would assume that the RPCs are also working as they weren't touched but please double check for me as I don't have a good test set up.

TasoOneAsia commented:
> Well that's good, appreciate the review. I've tested the net promise functionality (which was the whole original point of this PR) in a server and that's working fine. I would assume that the RPCs are also working as they weren't touched but please double check for me as I don't have a good test set up.

Yup can check, believe it or not when I started to get busy and left this project, I was working on creating a simple test runner that would work within a FXServer context but lets see if I can find what's left of that..

napei commented:
> Yup can check, believe it or not when I started to get busy and left this project, I was working on creating a simple test runner that would work within a FXServer context but lets see if I can find what's left of that..

That sounds interesting. Another repo to maintain! 😂

napei commented:
Noticed a couple more semantic things that could be moved around to make it a bit more readable in 8c93f02...f534042.

TasoOneAsia commented:
Yup looking through now. I am going to really try and get this PR merged in tonight with a new version published along side it.

TasoOneAsia reviewed (approved):
Good work :)

TasoOneAsia commented:
Testing didn't have any issues, gonna merge this to master but won't publish new version until a little bit later due to automation CI. Thanks again, heavily appreciate it.

TasoOneAsia commented:
Thanks for implicitly taking care of those package bumps as well, literally closed everything pending.

napei commented:
> Testing didn't have any issues, gonna merge this to master but won't publish new version until a little bit later due to automation CI. Thanks again, heavily appreciate it.

Awesome! No worries at all.

> Thanks for implicitly taking care of those package bumps as well, literally closed like 5 PRs

Yeah figured I might as well haha
